### ISSUE TITLE
Prevent Empty Checkpointable Data Structure Restores

### ISSUE THREAD

**Issue by sleighsoft** [tensorflow/tensorflow#23192]
@allenlavoie

Issue tracking the "bug" that is caused when restoring a checkpoint where the Checkpointable object has new checkpointable data structures that do not contain any checkpointable objects themselves and therefore prevent restoration when using `load_status.assert_consumed()`.

*Example Code:*
```python
import tensorflow as tf
from tensorflow.python.training.checkpointable.tracking import Checkpointable


class Model(Checkpointable):

  def __init__(self):
    self.variable = tf.get_variable("variable", [2, 2])
    self.dict = {
        "test": 1,
        "test2": 2,
    }
    self.dict_var = {
        "test": 1,
        "test2": tf.get_variable("dict_var", [2, 2])
    }
    self.dict_nested_var = {
        "test": 1,
        "test2": {
            "var": tf.get_variable("dict_nested_var", [2, 2])
        }
    }
    self.list_var = [tf.get_variable("list_var", [2, 2])]


print('SAVE')
with tf.Graph().as_default():

  s = tf.Session()

  m = Model()

  s.run(tf.global_variables_initializer())

  c = tf.train.Checkpoint(model=m)

  c.save('checkpoints/', session=s)

print('RESTORE')
with tf.Graph().as_default():

  s = tf.Session()

  m = Model()

  m.dict_new = {
      "test": 1,
      "test2": 2
  }

  c = tf.train.Checkpoint(model=m)

  status = c.restore('checkpoints/-1')

  status.assert_consumed().run_restore_ops(s)
```

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by wt-huang** [comment]
@sleighsoft When you write new objects in c.restore() which were not originally in the checkpointable data structure, it would not work. This is the intent of save/restore checkpoint functionality. 

If you plan to change the checkpoints, best go back to c.save() to make the change then call c.restore(). It is probably much more manageable and less confusion this way.

**Comment by allenlavoie** [comment]
I think the point is that auto-generated list/dict wrappers which don't have any variables or checkpointable objects associated with them shouldn't count as a checkpoint not matching. Otherwise `self._unused_list = []` would break an assertion, which is technically correct-ish but not helpful. I could have conditionals in my __init__ which only sometimes assign such an attribute, in which case small configuration changes would accidentally break assertions.

Either way the checkpoint restore works fine, since no variables were added or tracked in a different way. It's just a question of whether the optional assertions pass or not.

**Comment by sleighsoft** [comment]
What @allenlavoie says :)

**Comment by tensorflowbutler** [comment]
Nagging Assignees @allenlavoie, @wt-huang: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by mohantym** [comment]
Hi @sleighsoft !
Code in above template have depreciated  now. Could you please provide 2.x version code to replicate this issue.
Thank you!
